### ISSUE TITLE
False positive TypeDoesNotContainType  (trying to reconcile type 'non-empty-array<string, string>' to empty)

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#1297]
This occurs with the below snippet on psalm's master branch.

This seems to be related to using self::KEY_SOMETHING instead of 'z'. However, even if the value was a dynamic string, it still should not be emitted for `empty($arr[unknown offset])`

```php
<?php

class R {
    /** @var bool */
    public $requireSomething;
    /** @var array<string,string> */
    public $requiredFields;

    const E_SOMETHING_INVALID = 2;
    const KEY_SOMETHING = 'z';
    /** @param array<string,string> $requiredFields */
    public function __construct(array $requiredFields) {
        $this->requireSomething = rand(0,1) > 0;
        $this->requiredFields = $requiredFields;
    }
    /**
     * @param array<string,string> $data
     * @return array
     */
    public function validateInput(array $data) : array {
        $errors = [];
        foreach ($this->requiredFields as $key => $unused_required) {
            if (empty($data[$key])) {
                $errors[] = 4;
            }
        }

        if (!empty($data['otherfield'])) {
            // Observed: Psalm emits TypeDoesNotContainType - Found a contradiction when evaluating $data and trying to reconcile type 'non-empty-array<string, string>' to empty
            // Expected: Psalm does not emit false positive
            // NOTE: If self::KEY_SOMETHING is replaced by 'z', then psalm does not warn.
            if ($this->requireSomething && empty($data[self::KEY_SOMETHING])) {
                $errors[] = 2;
            }
        }

        return $data;
    }
}
```

**Comment by muglug** [comment]
Reproduced with https://getpsalm.org/r/1534a433a4
```php
class Foo {
  const BAR = 'bar';
}

/**
 * @param array<string,string> $data
 */
function bat(array $data) : void {
  if (!empty($data['foo'])) {
    if (empty($data[Foo::BAR])) {}
  }
}
```
